src/EFCore/ChangeTracking/DoubleValueComparer.cs

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq.Expressions;
+using JetBrains.Annotations;
+
+namespace Microsoft.EntityFrameworkCore.ChangeTracking
+{
+    /// <summary>
+    ///     Defines value comparison for <see cref="double" /> which correctly takes <see cref="double.NaN" />
+    ///     into account.
+    /// </summary>
+    public class DoubleValueComparer : ValueComparer<double>
+    {
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="DoubleValueComparer" /> class.
+        /// </summary>
+        public DoubleValueComparer() : base(
+            (x, y) => double.IsNaN(x) ? double.IsNaN(y) : x.Equals(y),
```

Comment:
I see it's merged now, but
```suggestion
            (x, y) => x.Equals(y),
```
will yield the same result.

Reply:
Thanks, did not know that... Do you want to submit a cleanup PR for this (and the same for FloatValueComparer)?

Reply:
Yep, will send one.

Reply:
--> https://github.com/dotnet/efcore/pull/22212

Reply:
If the value comparer are just doing Equals then do we need a custom value comparer?

Reply:
Agree with @smitpatel. What's the actual issue we're trying to fix here?

Reply:
Here's what's happening.

* The result of `double.NaN == double.NaN` is false, but the result of `double.NaN.Equals(double.NaN)` is true.
* Prevoious to this PR, floating point properties would get a [default ValueComparer](https://github.com/dotnet/efcore/blob/release/5.0/src/EFCore/ChangeTracking/ValueComparer.cs#L196), whose implementation of Equals is Expression.Equal. That calls the "operator" version which returns false.
* This PR (and #22212) fixes that by having a specialized value comparer which calls double.Equals instead of ==, which works correctly.

It may be better to fix this by having a DefaultDoubleValueComparer which is the same as DefaultValueComparer, but calls Equals instead of ==: it's cleaner and would work universally (e.g. Cosmos). @ajcvickers if that makes sense I'll submit that. If we don't do that, we probably want to clean up and remove double/float from DefaultValueComparer as it doesn't work correctly for NaN.

Reply:
@roji Yes, do that.

Reply:
@roji maybe it makes sense to close https://github.com/dotnet/efcore/pull/22212 and incorporate that change into your coming PR.

Reply:
Thanks @gfoidl, submitted #22242 to clean this up.